### PR TITLE
*: Add first version of the code

### PR DESCRIPTION
**Pull request by gaocegege**
/assign @ddysher 

```release-note
NONE
```

### CODE REVIEW

File: pkg/manager/manager.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+func New(Backend backend.Interface, S2IClient s2i.Interface, Interpreter interpreter.Interface) *Manager {
+	return &Manager{
+		Backend:     Backend,
+		S2IClient:   S2IClient,
+		Interpreter: Interpreter,
+	}
+}
+
+func (m Manager) Execute(code string) (*types.Job, error) {
```
Comment (ddysher) [inline pkg/manager/manager.go]: add godoc for public methods

File: pkg/interpreter/simple/simple.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+func New() *Interpreter {
+	return &Interpreter{
+		FrameworkPrefix: "%kubeflow framework=",
+		WorkerPrefix:    "%kubeflow worker=",
+		PSPrefix:        "%kubeflow ps=",
+	}
+}
+
+func (i Interpreter) Preprocess(code string) (*types.Parameter, error) {
```
Comment (ddysher) [inline pkg/interpreter/simple/simple.go]: ditto, add godoc for public methods

File: pkg/backend/kubeflow/kubeflow.go [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+			LabelSelector: GetLabelSelectorForJob(job),
+		})
+		if err != nil {
+			fmt.Printf("[kubeflow] Failed to get pods for the given job %s\n", job.Name)
+		}
+		if len(pods.Items) != job.PS+job.Worker {
+			fmt.Printf("[kubeflow] Wating for %d PS and %d workers\n", job.PS, job.Worker)
+			time.Sleep(queryInterval)
+
+			if retry == retryTime {
```
Comment (ddysher) [inline pkg/backend/kubeflow/kubeflow.go]: nit: there's quite a few places for this retry logic, better to rewrite it into a function util (in coming PR)

File: pkg/backend/kubeflow/kubeflow.go [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+	logOpts := &v1.PodLogOptions{
+		Follow: true,
+	}
+	instanceName := GetReplicaInstanceForPod(job, pod)
+
+	PodRef := &pod
+
+	// Check if the pod is ready.
+	for {
+		if PodRef.Status.Phase == v1.PodPending {
```
Comment (ddysher) [inline pkg/backend/kubeflow/kubeflow.go]: notes for improvement (future PR): for better ux, we'd better provide error information to user as well, e.g. if pending for a long time, try return event message.

File: pkg/backend/kubeflow/kubeflow.go [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+			continue
+		} else {
+			break
+		}
+	}
+
+	fmt.Printf("[kubeflow] There are %d pods for the job.\n", len(pods.Items))
+
+	wg.Add(len(pods.Items))
+	for _, pod := range pods.Items {
```
Comment (ddysher) [inline pkg/backend/kubeflow/kubeflow.go]: This can be messy .... let's try come up with a better design later.

File: scripts/hack_install.sh [hunk 6]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
```
Comment (ddysher) [inline scripts/hack_install.sh]: generally we just use `./hack/install.sh`

File: pkg/types/types.go [hunk 7]
```diff
@@ -0,0 +1,23 @@
+package types
+
+// TODO: Add more configs.
+type Parameter struct {
```
Comment (ddysher) [inline pkg/types/types.go]: is this user facing?  if so, let's make it a versioned package?

File: pkg/interpreter/simple/simple.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/caicloud/ciao/pkg/types"
+)
+
+type Interpreter struct {
```
Comment (ddysher) [inline pkg/interpreter/simple/simple.go]: note for improvement (future PR)

This is the core API of the kernel but it's nested too deeply, I would imagine this to be easily discoverable. The specific commands should be defined somewhere else and preferably versioned, and interpreter just imports the commands?  This is also easier for us to implement features like list all available commands, etc

File: pkg/interpreter/simple/simple.go [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+	} else if strings.Contains(line, i.PSPrefix) {
+		param.PSCount, err = strconv.Atoi(line[len(i.PSPrefix):])
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (i Interpreter) PreprocessedCode(code string) string {
```
Comment (ddysher) [inline pkg/interpreter/simple/simple.go]: ditto, add godoc for public methods

File: pkg/kernel/kernel.go [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+	}()
+
+	go func() {
+		defer writersWG.Done()
+		jupyterStdErr := JupyterStreamWriter{StreamStderr, &receipt}
+		io.Copy(&jupyterStdErr, rErr)
+	}()
+
+	// eval
+	vals, executionErr := k.doEval(code)
```
Comment (ddysher) [inline pkg/kernel/kernel.go]: As discussed offline, this should pass code to ipython if there's no kubeflow magic command?

File: artifacts/kernel.json [hunk 11]
```diff
@@ -0,0 +1,8 @@
+{
+  "argv": [
+    "kubeflow-kernel",
+    "{connection_file}"
+  ],
+  "display_name": "Kubeflow",
```
Comment (ddysher) [inline artifacts/kernel.json]: s/Kubeflow/Kubeflow-PythonN?

File: artifacts/kernel.json [hunk 12]
```diff
@@ -0,0 +1,8 @@
...
+  "language": "Python"
```
Comment (ddysher) [inline artifacts/kernel.json]: Python3 vs Python2?  we probably need to make clear (even though the pr doesn't implement python integration)

File: docs/demo.md [hunk 13]
```diff
@@ -0,0 +1,314 @@
+# Demo Show
+
+In this demo, we will show how to create a distributed traning in Jupyter Notebook with the help of Ciao.
+
+[Video on Vimeo](https://vimeo.com/288833803)
+
+The code in the cell is:
+
+```python
```
Comment (ddysher) [inline docs/demo.md]: note for future improvement

The user experience here is that they essentially have to copy/paste their already well-written code and submit to kubeflow, i.e. mostly the training part.  What I have in mind is that at last, we provide a more smoother experience such that it's possible to perform "develop-train-eval".

File: version/version.go [hunk 14]
```diff
@@ -0,0 +1,9 @@
+package version
+
+const (
+	// Version defines the gophernotes version.
+	Version string = "1.0.0"
+
+	// ProtocolVersion defines the Jupyter protocol version.
+	ProtocolVersion string = "5.0"
```
Comment (ddysher) [inline version/version.go]: s/ProtocolVersion/JupyterProtocolVersion?

File: cmd/kubeflow-kernel/main.go [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+	// 	kubeconfig = os.Getenv(RecommendedKubeConfigPathEnv)
+	// }
+
+	// Get kubernetes config.
+	kcfg, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		log.Fatalf("Error building kubeconfig: %s", err.Error())
+	}
+
+	kubeflowBackend, err := kubeflowbackend.New(kcfg)
```
[the remark this answers is not in the document]
Comment (gaocegege) [reply cmd/kubeflow-kernel/main.go]: In my opinion, we will define a config for the kernel, to specify using kubeflow or other backend, s2i or knative.

File: artifacts/kernel.json (hunk 11 above)
Comment (gaocegege) [reply artifacts/kernel.json]: I think we will support Python 2 & 3 in the future, while now we can set to Kubeflow-Python3

File: pkg/kernel/kernel.go (hunk 10 above)
Comment (gaocegege) [reply pkg/kernel/kernel.go]: We could discuss about it in #8

File: version/version.go (hunk 14 above)
Comment (gaocegege) [reply version/version.go]: I prefer to keep ProtocolVersion since it is copied from gophernotes.

File: pkg/types/types.go (hunk 7 above)
Comment (gaocegege) [reply pkg/types/types.go]: It is not user-facing. the magic commands are user facing and I think we should make that a versioned package in the future (After we have a real interpreter.)

**Comment by gaocegege** [comment]
@ddysher Addressed, PTAL

**Comment by ddysher** [comment]
Thanks! btw, can u add `vendor`?

/lgtm

**Comment by ddysher** [comment]
First PR without OWNERS file, manually merge.

**Pull request merged**